### PR TITLE
fix(store): reject empty sync fields

### LINKED ISSUE #686: fix(sync): empty prompt content accepted at capture poisons cloud push for all projects; repair cannot fix prompt entities
### 📋 Pre-flight Checks

- [x] I have searched [existing issues](https://github.com/Gentleman-Programming/engram/issues) and this is not a duplicate
- [x] I understand this issue needs `status:approved` before a PR can be opened

### 📝 Bug Description

Prompt capture accepts empty `content`, but the cloud server rejects that same payload at push time with `500: prompt payload content is required for upsert`. Because the push queue is sequential and global across projects, a single empty-prompt mutation permanently stalls cloud push for **every** project until repaired by hand. In our case one empty prompt captured on 2026-07-22 silently blocked ~500 pending mutations across 4 projects for 8 days (174 consecutive failures), while pull kept working — so devices that were already mirrored reported a healthy sync and the data loss was invisible from their side.

Three gaps combine into this:

1. **No capture-time validation.** `handleSavePrompt` (`internal/mcp/mcp.go:1564`) and `Store.AddPrompt` (`internal/store/store.go:2549`) accept and persist `content=""` verbatim, and enqueue the upsert mutation. The server enforces `content` as required (cloudserver/chunkcodec canonicalization), so the mutation is un-pushable from the moment it is created. This is the prompt-entity mirror of #459 (empty observation titles).
2. **No poison-mutation handling in the push queue.** The failing mutation is retried forever and everything behind it — including other projects' mutations — waits indefinitely. Same failure family as the import-side stalls in #649 / #671, but on the push side.
3. **Repair does not support the prompt entity.** `engram cloud upgrade repair` reports `manual-action-required` and offers nothing; `docs/engram-cloud/troubleshooting.md` documents loop/interactive repair for `entity=session|observation op=upsert` only. Relates to #340 (required-field repair), which lists session/observation blocker classes but not prompts.

### 🔄 Steps to Reproduce

1. Configure a cloud target and enroll a project (`engram cloud enroll <project>`).
2. Save a prompt with empty content (e.g. `mem_save_prompt` with `content: ""` — in our case it came from automated hook capture): a `user_prompts` row and an `upsert` mutation with `"content":""` are created.
3. Let the daemon push (or run `engram sync --cloud --project <project>`).
4. Push fails: `mutation push: status 500: insert mutations: cloudstore: canonicalize materialized mutation batch chunk: mutations[N]: prompt payload content is required for upsert`.
5. `sync_state` goes `degraded`, `consecutive_failures` grows unbounded; mutations from **all** projects behind that seq stop replicating.
6. `engram doctor` → `sync_mutation_payload_missing_required_fields` (correct detection).
7. `engram cloud upgrade doctor/repair --project <project>` → `upgrade_blocked_legacy_mutation_manual`, `manual-action-required: prompt payload missing required upsert fields: content` — no repair path offered.

Recovery required manual surgery: `engram delete prompt <id>` (to tombstone the source row) plus a hand-written `DELETE FROM sync_mutations WHERE seq=<N>` in SQLite.

### ✅ Expected Behavior

Any of these layers would have prevented or contained it (ordered by depth):

1. **Capture:** reject or skip persisting/enqueueing prompts whose content is empty after normalization — same principle as #459.
2. **Queue:** quarantine a mutation the server permanently rejects (validation-class 500) into a dead-letter state visible in `doctor`, instead of retrying forever and blocking unrelated projects.
3. **Repair:** extend `cloud upgrade repair` to the `prompt` entity (drop-or-tombstone flow for empty content), so no manual SQL is needed.

### ❌ Actual Behavior

Empty prompt persisted and enqueued; push retries the same mutation indefinitely (174 consecutive failures over 8 days); all projects' cloud replication stalls; repair declines with `manual-action-required`; recovery requires manual SQLite edits.

### Operating System

Linux (Other)

### Engram Version

1.20.0

### Agent / Client

Claude Code

### Relevant Logs

```
Sync diagnostic: degraded
reason_code: transport_failed
reason_message: push: transport push project "owlvision": cloud: mutation push: status 500: insert mutations: cloudstore: canonicalize materialized mutation batch chunk: mutations[1]: prompt payload content is required for upsert

$ engram doctor
[blocked] sync_mutation_required_fields — 1 finding(s) detected.
  - sync_mutation_payload_missing_required_fields: prompt payload missing required fields: content
    evidence: {"entity":"prompt","entity_key":"prompt-a1fdc6ceb6b624fb","missing_fields":["content"],"op":"upsert","project":"owlvision","seq":7783,"target_key":"cloud"}

$ engram cloud upgrade repair --project owlvision --dry-run
class: blocked
reason_code: upgrade_blocked_legacy_mutation_manual
message: manual-action-required: prompt payload missing required upsert fields: content (seq=7783 entity=prompt entity_key="prompt-a1fdc6ceb6b624fb" op=upsert)
applied: false

sync_state (before manual repair): cloud | degraded | last_enqueued_seq=8283 | last_acked_seq=7781 | consecutive_failures=174
```

### Additional Context

- The stored mutation payload was `{"sync_id":"prompt-a1fdc6ceb6b624fb","session_id":"...","content":"","project":"owlvision","created_at":"2026-07-22 13:43:09"}` and the source `user_prompts` row had `content=''` as well — so this is not payload drift; the empty value was accepted at capture.
- Related: #459 (capture-time validation for observation titles), #340 (required-field repair, session/observation classes), #649 / #671 (import-side poison-mutation stalls), #630 (source row re-emits broken mutation after payload-only repair).
- Happy to contribute a PR for the capture-time validation slice once approved.

### LINKED ISSUE #695: bug(mcp): mem_save does not enforce its required title; a blank one permanently stalls cloud sync for every project
### 📋 Pre-flight Checks

- [x] I have searched existing issues and this is not a duplicate
- [x] I understand this issue needs `status:approved` before a PR can be opened

### 📝 Bug Description

`mem_save` declares `title` as a required argument but never enforces it. When a client omits it, the observation is stored with `title = ""`. That row then produces a sync mutation the cloud canonicalizer rejects permanently, and because the outbound push is batched, **it stalls the queue for every project on the machine** — silently, and with no way to repair it afterwards.

The schema says required — `internal/mcp/mcp.go:333-335`:

```go
mcp.WithString("title",
    mcp.Required(),
    mcp.Description("Short, searchable title (e.g. 'JWT auth middleware', 'Fixed N+1 query')"),
),
```

The handler does not check it — `internal/mcp/mcp.go:1186-1195`:

```go
func handleSave(s *store.Store, cfg MCPConfig, activity *SessionActivity) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		title, _ := req.GetArguments()["title"].(string)   // ← failed assertion discarded
		content, _ := req.GetArguments()["content"].(string)
		...
		if strings.TrimSpace(content) == "" {
			return mcp.NewToolResultError("content is required for mem_save (...)"), nil
		}
```

The asymmetry is the tell: `content` has an explicit guard one line below, `title` has none. The `_` swallows the missing argument and yields `""`.

The HTTP path is not affected — `internal/server/server.go:330` rejects a blank title — so this is specific to the MCP entry point, which is how agents actually write.

### Why it is not merely cosmetic

Downstream, an empty title is unrecoverable:

1. The cloud canonicalizer requires a non-empty title for an observation upsert, so the mutation is rejected for the lifetime of the queue.
2. `engram cloud upgrade repair` cannot fix it. It fills missing payload fields from the local table (`internal/store/store.go:1594-1600`) but blocks when the local value is also empty (`:1613-1624`) — which it always is, because that is where the blank came from.
3. `mem_doctor` reports `status: ok` throughout.

### 🔄 Steps to Reproduce

1. Call `mem_save` through MCP with `content` and `type` but no `title`.
2. Confirm the stored row: `SELECT title FROM observations WHERE id = <id>` → `''`.
3. With cloud sync configured, observe the outbound queue stop draining. `engram cloud upgrade doctor --project <P>` reports `upgrade_blocked_legacy_mutation_manual` / `observation payload missing required upsert fields: title`.

### ✅ Expected Behavior

`mem_save` should never persist a blank title — either by rejecting the call the way `content` is rejected, or by deriving one from the body.

### ❌ Actual Behavior

The save reports success, the row is stored with `title = ""`, and cloud sync for every project stops.

### Observed impact

On one installation, an SDD phase began saving without `title`. Measured from `sync_mutations.acked_at`:

- 17,343 mutations acked between 2026-06-28 and 2026-07-28 11:35:21 — sync was healthy.
- First blank-title observation created 2026-07-28 13:28:58, under two hours later.
- Nothing synced again. 8 days, **2,682 mutations pending across 12 projects**, 44 of them blank-title observations. No error surfaced anywhere.

### Engram Version

`1.20.0`

### Operating System

Linux (Ubuntu/Debian)

### Agent / Client

Claude Code

### 💡 Proposed fix

Rejecting would be truest to the declared schema, but it silently loses a memory an agent believed it had saved. Deriving keeps the data and still guarantees the invariant, so that is what I implemented and verified — placed in `store.AddObservation`/`UpdateObservation` rather than the MCP handler so every write path is covered:

```go
// An observation stored with a blank title yields a sync mutation the cloud
// canonicalizer rejects permanently, which stalls the outbound queue for
// every project. Never persist one: derive a title from the body instead.
if title == "" {
	title = DeriveTitleFromContent(content)
}
```

`DeriveTitleFromContent` takes the first non-empty line and strips the markdown decoration agents habitually put there (headings, bullets, a leading `**What**:` key, code ticks), capping the result. The cap cuts on a rune boundary — `len()` is bytes and observation bodies are routinely non-ASCII.

Verified on a patched `v1.20.0` tree:

- `go build ./...` and `go vet ./...` clean; `go test ./...` exits 0, 22 packages, no failures.
- New tests: table cases for the derivation, a rune-boundary case, and `TestAddObservationNeverStoresBlankTitle` asserting the invariant. That last one fails on unpatched code with `observation stored with a blank title; this poisons the sync queue`, and passes with the change.
- End to end on a real installation: `mem_save` through MCP with no `title`, body `## FIXA-PROBE sin titulo`, stored as `FIXA-PROBE sin titulo`.

Happy to open the PR once this has `status:approved`. If you would rather reject than derive, say so and I will submit that shape instead — the invariant matters more than which way it is enforced.

Filed separately: the reason one such row can take down every project is a distinct defect in the push loop, which I have written up on its own.

### PR DESCRIPTION
**Pull request by dnlrsls**
## Linked Issues

Closes #686
Closes #695

---

## PR Type

- [x] `type:bug` - Bug fix
- [ ] `type:feature` - New feature
- [ ] `type:docs` - Documentation only
- [ ] `type:refactor` - Code refactoring (no behavior change)
- [ ] `type:chore` - Maintenance, dependencies, tooling
- [ ] `type:breaking-change` - Breaking change

---

## Summary

- Rejects blank normalized observation titles and prompt content at the Store admission boundary.
- Keeps invalid writes from reaching persistence or the sync journal.
- Returns HTTP 400 for these client admission errors while preserving existing mappings for other failures.

## Changes

| File | Change |
|------|--------|
| `internal/store/store.go` | Add stable admission errors and validate required fields before transactions. |
| `internal/store/store_test.go` | Prove invalid create, upsert, update, and prompt writes leave rows and sync mutations unchanged, including `inserted == false`. |
| `internal/mcp/mcp_test.go` | Prove MCP handlers expose Store admission failures without persistence. |
| `internal/server/server.go` | Map Store admission errors to HTTP 400. |
| `internal/server/server_test.go` | Prove whitespace-only writes return 400 and do not persist. |

## Test Plan

- [x] Focused Store admission tests
- [x] `go test ./internal/mcp -run TestSaveHandlersExposeStoreAdmissionErrors -count=1`
- [x] `go test ./internal/mcp`
- [x] `go test ./internal/server -count=1`
- [x] `go vet ./internal/server ./internal/store`
- [x] `gofmt` and `git diff --check`
- [ ] `go test ./internal/store -count=1` - blocked by pre-existing Windows TempDir cleanup failures in `TestMigrate_Idempotent` and `TestNewErrorBranches`

---

## Contributor Checklist

- [x] I linked approved issues above (`Closes #686`, `Closes #695`)
- [x] I added exactly one `type:*` label to this PR
- [x] I ran relevant unit tests locally
- [ ] I ran the full E2E suite locally
- [x] Docs updated (not required; no public contract changed)
- [x] Commits follow conventional commits format
- [x] No `Co-Authored-By` trailers in commits

---

## Notes for Reviewers

Validation lives in Store so every live-write caller receives the same canonical invariant. Import, repair, and cloud-apply semantics are intentionally unchanged.

Native RDD completed an independent four-lens review and approved the exact committed tree `4138cd52bd8254d5a2000784c78da31072181fa6` without blocking findings. Delivery remains governed by ordinary repository policy.

## Summary by CodeRabbit

- **Bug Fixes**
  - Added validation to prevent observations with blank titles or content from being created or updated.
  - Added validation to prevent prompts with blank content from being created.
  - Invalid submissions now return a clear client error instead of an internal server error.
  - Rejected changes no longer create or modify stored records.
- **Tests**
  - Expanded coverage for whitespace-only fields across observation and prompt workflows.

### CODE REVIEW

File: internal/store/store_test.go [hunk 1]
```diff
@@ -114,6 +114,165 @@ func TestAddObservationDeduplicatesWithinWindow(t *testing.T) {
...
+				_, err := s.AddPrompt(AddPromptParams{SessionID: "s-prompt-admission", Content: " \t\n ", Project: "engram"})
+				return err
+			},
+		},
+		{
+			name: "add prompt if missing",
+			add: func(s *Store) error {
+				_, _, err := s.AddPromptIfMissing(AddPromptParams{SessionID: "s-prompt-admission", Content: " \t\n ", Project: "engram"})
+				return err
+			},
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply internal/store/store_test.go]: Fixed in 1113d6e. The test now captures the `inserted` result from `AddPromptIfMissing` and asserts it remains false alongside the sentinel-error and no-persistence checks.

File: internal/store/store.go [hunk 2]
```diff
@@ -43,14 +43,16 @@ var sqliteWriteRetryBackoffs = []time.Duration{
...
-	ErrPromptNotFound         = errors.New("prompt not found")
-	ErrProjectNotFound        = errors.New("project not found")
+	ErrSessionNotFound          = errors.New("session not found")
+	ErrSessionHasObservations   = errors.New("session still has observations")
+	ErrSessionDeleteBlocked     = errors.New("session deletion is blocked while cloud sync enrollment is active")
+	ErrObservationNotFound      = errors.New("observation not found")
+	ErrPromptNotFound           = errors.New("prompt not found")
+	ErrProjectNotFound          = errors.New("project not found")
+	ErrObservationTitleRequired = errors.New("observation title is required")
+	ErrPromptContentRequired    = errors.New("prompt content is required")
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply internal/store/store.go]: Fixed in 1113d6e. The HTTP handlers now use `errors.Is` to map the title and prompt admission errors to 400 while preserving every other existing error mapping. The server regression also verifies that the rejected writes do not persist.

**Review by Alan-TheGentleman — APPROVED** [review]
Verified in depth at the original head 1113d6ea and extended at head 39d9f975 to close the two remaining gaps in the same poison class: blank observation content is now rejected at admission in AddObservation and UpdateObservation with ErrObservationContentRequired, mapped to 400 in both HTTP handlers, mirroring the PR's title and prompt guards before any transaction. Caller audit found no legitimate blank-content writer, import and sync-apply paths stay untouched by design, and the tests prove zero row and zero sync_mutations deltas plus seeded-row immutability for every rejected shape. Both #686 and #695 are still live on main and this closes their capture layer. Merging. The queue-quarantine and prompt-repair layers #686 listed remain follow-up work now that #766 covers the blast radius.
